Framework/Core/src/ServiceRegistry.cxx

```diff
@@ -0,0 +1,63 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include "Framework/ServiceRegistry.h"
+#include "Framework/Tracing.h"
+#include "Framework/Logger.h"
+#include <iostream>
+
+namespace o2::framework
+{
+
+ServiceRegistryBase::ServiceRegistryBase()
+{
+  mServicesKey.fill(0L);
+  mServicesValue.fill(nullptr);
+  for (size_t i = 0; i < mServicesBooked.size(); ++i) {
+    mServicesBooked[i] = false;
+  }
+}
+
+/// Type erased service registration. @a typeHash is the
+/// hash used to identify the service, @a service is
+/// a type erased pointer to the service itself.
+/// This method is supposed to be thread safe
+void ServiceRegistryBase::registerService(hash_type typeHash, void* service, ServiceKind kind, uint64_t threadId, const char* name) const
+{
+  hash_type id = typeHash & MAX_SERVICES_MASK;
+  hash_type threadHashId = (typeHash ^ threadId) & MAX_SERVICES_MASK;
+  // If kind is not stream, there is only one copy of our service.
+  // So we look if it is already registered and reused it if it is.
+  // If not, we register it as thread id 0 and as the passed one.
+  if (kind != ServiceKind::Stream && threadId != 0) {
+    void* oldService = this->get(typeHash, 0, kind);
+    if (oldService == nullptr) {
+      registerService(typeHash, service, kind, 0);
+    } else {
+      service = oldService;
+    }
+  }
+  for (uint8_t i = 0; i < MAX_DISTANCE; ++i) {
+    // If the service slot was not taken, take it atomically
+    bool expected = false;
+    if (mServicesBooked[i + threadHashId].compare_exchange_strong(expected, true,
+                                                                  std::memory_order_seq_cst)) {
+      mServicesValue[i + threadHashId] = service;
+      mServicesMeta[i + threadHashId] = ServiceMeta{kind, threadId};
+      mServicesKey[i + threadHashId] = typeHash;
+      std::atomic_thread_fence(std::memory_order_release);
+      return;
+    }
```

Comment:
@ktf : Without understanding 100% what is happening: The `mServicesBooked` is obviously atomic, so only a single thread will ever write to the same `mServicesValue[x]` / `mServicesKey[x]`.
However, since there is no fence in between and the variables are not volatile, I believe the compiler may reorder the write to mServicesValue and mServicesKey, so basically the reader could read the key before the values was written. (But not 100% sure. I believe the CPU must not reorder, but the compiler could.

I don't understand what you mean by "the partial write is taken back" or where this should happen.

Reply:
not sure if the compare_exchange does not release already throwing this out of whack...

Reply:
@davidrohr reading the value before the key was written in the separate thread should be prevented by the `std::atomic_thread_fence(std::memory_order_acquire)` I've in the get, no?

Reply:
@mkrzewic compare_exchange will do the release for the atomic itself, but not for the rest of the stuff I manipulate in the if, no?

Reply:
Not sure, I think in that case you'd need the fence between the two not afterwards.

Reply:
> not sure if the compare_exchange does not release already throwing this out of whack...

I think the compare exchange is not the point. It is only used for the filling thread, to synchronize that not 2 threads write to the same entry.
The reading thread never checks mServicesBooked but only mServicesKey, if I understand correctly. But yeah, that confused me as well.

Reply:
@davidrohr I think on some architectures what might happen is that a speculative store op is issue before the check for the atomic and committed only once the check if verified to be true. A separate (hyper)thread might see the not-yet-commited store, and misbehave. But maybe that's not the case on x86...

Reply:
I think Intel has this transactional memory feature, but it must be used explicitly. But anyway, I don't believe we have to think on that hardware level here, but just stick to C++ specs, since you only use C++ fence features.

I think the fully correct way of doing it is:
make mServicesKey atomic<int>

write:
- write mServicesValue
- memory fence (release)
- mServicesKey[i].atomic_store(key)

read:
- key = mServicesKey[i].atomic_load()
- check key
- memory fence (acquire)
- read mServicesValue

although I am not fully sure if making mServicesKey an atomic is required. It is an integer, so stores and loads should anyway be atomic.

Reply:
when in doubt: explicit mutex :)

Reply:
@davidrohr I think what I have here is the same, just with the Booked vector, rather than the key. So I think we are on the same page. What I am after is the behaviour described by the last sentence of (citing cppreference):

> A release fence F in thread A synchronizes-with atomic acquire operation Y in thread B, if
there exists an atomic store X (with any memory order)
> Y reads the value written by X (or the value would be written by release sequence headed by X if X were a release operation)
> X is sequenced-before F in thread A
> In this case, all non-atomic and relaxed atomic stores that are sequenced-before F in thread A will happen-before all non-atomic and relaxed atomic loads from the same locations made in thread B after Y.